test/extended/operators/olm.go

```diff
@@ -87,4 +91,66 @@ var _ = g.Describe("[Feature:Platform] OLM should", func() {
 			o.Expect(msg).To(o.Equal("IfNotPresent"))
 		}
 	})
+
+	// OCP-22070 author: jiazha@redhat.com
+	g.It("[ocp-22070]support grpc sourcetype", func() {
+		buildPruningBaseDir := exutil.FixturePath("testdata", "olm")
+		catalogsourceYAML := filepath.Join(buildPruningBaseDir, "image-catalogsource.yaml")
+		subscriptionYAML := filepath.Join(buildPruningBaseDir, "image-sub.yaml")
+
+		// Check packagemanifest: Test Operators
+		_, err := oc.AsAdmin().WithoutNamespace().Run("create").Args("-f", catalogsourceYAML).Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		err = wait.Poll(3*time.Second, 120*time.Second, func() (bool, error) {
+			output, err := oc.AsAdmin().WithoutNamespace().Run("get").Args("packagemanifest").Output()
+			if err != nil {
+				e2e.Logf("Fail to get packagemanifest, error:%v", err)
+				return false, nil
+			}
+			if strings.Contains(output, "Test Operators") {
+				return true, nil
+			}
+			return false, nil
+		})
+
+		// Check the Subscription, InstallPlan, CSV
+		_, err = oc.AsAdmin().WithoutNamespace().Run("create").Args("-f", subscriptionYAML).Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		getCmds := map[int][]string{
+			0: {"subscription", "test-operator"},
+			1: {"clusterserviceversion", "etcdoperator.v0.9.2"},
+			2: {"pods", "etcd-operator"},
+		}
+		var keys []int
+		for k := range getCmds {
+			keys = append(keys, k)
+		}
+		sort.Ints(keys)
```

Comment:
I can't see a reason to use a map here. It adds a lot of extra code too. A simple alternative would be to use a slice of slices:

```go
getCmds := [][]string{
    {"subscription", "test-operator"},
    {"clusterserviceversion", "etcdoperator.v0.9.2"},
    {"pods", "etcd-operator"},
}

for _, cmd := range getCmds {
    err := wait.Poll(3*time.Second, 120*time.Second, func() (bool, error) {
        output, err := oc.AsAdmin().WithoutNamespace().Run("get").Args(cmd[0], "-n", "openshift-operators").Output()
        if err != nil {
	    e2e.Logf("Fail to get %v, error:%v", cmd[0], err)
	    return false, nil
        }
        if strings.Contains(output, cmd[1]) {
            return true, nil
        }
        return false, nil
    })
    o.Expect(err).NotTo(o.HaveOccurred())
}
```

If you want something more foolproof, you can use a slice of structs (what I **strongly** recommend):

```go
getCmds := []struct{
    kind
    expect
}{
    {"subscription", "test-operator"},
    {"clusterserviceversion", "etcdoperator.v0.9.2"},
    {"pods", "etcd-operator"},
}

for _, cmd := range getCmds {
    err := wait.Poll(3*time.Second, 120*time.Second, func() (bool, error) {
        output, err := oc.AsAdmin().WithoutNamespace().Run("get").Args(cmd.kind, "-n", "openshift-operators").Output()
        if err != nil {
	    e2e.Logf("Fail to get %v, error:%v", cmd.kind, err)
	    return false, nil
        }
        if strings.Contains(output, cmd.expect) {
            return true, nil
        }
        return false, nil
    })
    o.Expect(err).NotTo(o.HaveOccurred())
}
```

Reply:
Thanks very much! Updated it.